File: src/main/java/editor/gui/ccp/handler/CategoryTransferHandler.java
package editor.gui.ccp.handler;

import java.awt.datatransfer.Transferable;
import java.util.function.Consumer;
import java.util.function.Predicate;
import java.util.function.Supplier;

import javax.swing.JComponent;
import javax.swing.TransferHandler;

import editor.collection.deck.CategorySpec;
import editor.gui.ccp.data.CategoryTransferData;

/**
 * Handler for transferring category specifications between components in the handler.
 * 
 * @author Alec Roelke
 */
@SuppressWarnings("serial")
public class CategoryTransferHandler extends EditorTransferHandler
{
    /** Function supplying the category specification to transfer. */
    private final Supplier<CategorySpec> supplier;
    /** Function specifying how to remove cut data. */
    private final Consumer<CategorySpec> remove;

    /**
     * Create a new category transfer handler.
     * 
     * @param s function supplying the category specification to transfer
     * @param c function determining if the source component contains the category,
     * if applicable (should simply return false if not applicable)
     * @param a function indicating how to add the category to the source component
     * @param r function indicating how to remove the category if it is cut from the
     * source component
     */
    public CategoryTransferHandler(Supplier<CategorySpec> s, Predicate<CategorySpec> c, Predicate<CategorySpec> a, Consumer<CategorySpec> r)
    {
        super(new CategoryImportHandler(c, a));
        supplier = s;
        remove = r;
    }

    @Override
    public int getSourceActions(JComponent c)
    {
        return TransferHandler.MOVE | TransferHandler.COPY;
    }

    @Override
    public Transferable createTransferable(JComponent c)
    {
        return new CategoryTransferData(supplier.get());
    }

    @Override
    public void exportDone(JComponent source, Transferable data, int action)
    {
        if (action == TransferHandler.MOVE)
            remove.accept(((CategoryTransferData)data).spec);
    }
}

File: src/main/scala/editor/gui/ccp/handler/CategoryTransferHandler.scala
package editor.gui.ccp.handler

import editor.collection.Categorization
import editor.gui.ccp.data.CategoryTransferData

import java.awt.Component
import java.awt.datatransfer.Transferable
import javax.swing.JComponent
import javax.swing.TransferHandler

/**
 * Handler for transferring [[Categorization]]s beween elements.
 * 
 * @constructor create a new category transfer handler with functions indicating how to transfer components
 * @param supplier function supplying the [[Categorization]] to transfer
 * @param contains predicate indicating if a category is already contained by the target element
 * @param add function specifying how to add the [[Categorization]] to the target element
 * @param remove function specifying how to remove the [[Categorization]] from the source element if it is to be moved rather than copied
 * 
 * @author Alec Roelke
 */
class CategoryTransferHandler(supplier: () => Categorization, contains: (Categorization) => Boolean, add: (Categorization) => Boolean, remove: (Categorization) => Unit)
    extends EditorTransferHandler(Seq(CategoryImportHandler(contains, add))) {
  override def getSourceActions(c: JComponent) = TransferHandler.COPY_OR_MOVE
  override def createTransferable(c: JComponent) = CategoryTransferData(supplier())
  override def exportDone(source: JComponent, data: Transferable, action: Int) = if (action == TransferHandler.MOVE) data match {
    case c: CategoryTransferData => remove(c.spec)
    case x => throw ClassCastException(s"expected ${classOf[CategoryTransferData]}, got ${x.getClass}")
  }
}
